protocol/x/listing/keeper/listing.go

```diff
@@ -152,6 +153,43 @@ func (k Keeper) CreatePerpetual(
 	return perpetual.GetId(), nil
 }
 
+func (k Keeper) UpgradeIsolatedPerpetualToCross(
+	ctx sdk.Context,
+	perpetualId uint32,
+) error {
+	// Validate perpetual exists and is in isolated mode
+	perpetual, err := k.PerpetualsKeeper.GetPerpetual(ctx, perpetualId)
+	if err != nil {
+		return err
+	}
+	if perpetual.Params.GetMarketType() != perpetualtypes.PerpetualMarketType_PERPETUAL_MARKET_TYPE_ISOLATED {
+		return fmt.Errorf("perpetual %d is not an isolated perpetual and cannot be upgraded to cross", perpetualId)
+	}
+
+	err = k.SubaccountsKeeper.TransferIsolatedInsuranceFundToCross(ctx, perpetualId)
+	if err != nil {
+		return err
+	}
+
+	err = k.SubaccountsKeeper.TransferIsolatedCollateralToCross(ctx, perpetualId)
+	if err != nil {
+		return err
+	}
+
+	_, err = k.PerpetualsKeeper.SetPerpetualMarketType(
+		ctx,
+		perpetualId,
+		perpetualtypes.PerpetualMarketType_PERPETUAL_MARKET_TYPE_CROSS,
+	)
+	if err != nil {
+		return err
+	}
+
+	// TODO Propagate changes to indexer
```

Comment:
I would add the indexer event in the SetPerpetualMarketType function itself

Reply:
k that'll happen in a future PR